### ISSUE TITLE
dhcp: support custom DHCP Request options

### ISSUE THREAD

**Issue by SilverBut** [containernetworking/plugins#660]
Our production environment needs v4/v6 dual stack (and v6-only network), and we are using DHCP to manage address for different server/services/users. To get a address normally, DHCP `option 82`(https://www.rfc-editor.org/rfc/rfc3046.html) is required, which should be sent during DHCP-Request.

I'm writing code for this support, so in case of anyone else also writing code for this, please tell me if I can be any help.

**Comment by SilverBut** [comment]
Also, #614 and #640 is related with processing received DHCP options. So maybe I can work on these two issue later, if possible. 

Besides, the new config syntax is required to hold those settings. I'm planning to use this as the config, which should be compatible with previous versions: 

```json
{
    "cniVersion": "0.4.0",
    "name": "example",
    "plugins": [
        {
            "type": "macvlan",
            "master": "bond0.150",
            "ipam": {
                "type": "dhcp",
                // use option name from dhcp-options(5)
                "request": [
                    {
                        "default": true // TODO: add some predefined field list which we can process
                    },
                    {
                        "option": "domain-search",
                        "required": false // by default false. if true, the lease is ignored, and wait for correct one
                    }
                ],
                "send": [
                    {
                        "default": false
                    },
                    {
                        "option": "agent.circuit-id",
                        "value": "blacksteel.cluster.server:bond0/vlan150"
                    },
                    {
                        "option": "agent.remote-id",
                        "value-source": "container-id" // possible values: "container-id", "net-ns", "if-name"
                    }
                ]
            }
        }
    ]
}
```

**Comment by vivekskrishna** [comment]
Hi @SilverBut , have you made any progress in implementing this. i am happy to try this out.

**Comment by SilverBut** [comment]
@vivekskrishna Hi. Not ready for release since it does not work as intended in my own server. I'm still debugging...

**Comment by vivekskrishna** [comment]
@SilverBut Thanks for the update. Let me know if I can help in some way.

**Comment by dcbw** [comment]
I think these would be a good improvement for the DHCP plugin. Thanks for looking into it!

**Comment by SilverBut** [comment]
To make everyone know current status here is a brief report of my current work progress.

## Use argument from CNI arguments

As you have known, container name and some other arguments are not include in CNI (see #435), so for IPAM plugin, the only possible way to get is (hope) runtime can set correct value in [`CNI_ARGS`](https://github.com/containernetworking/cni/blob/master/SPEC.md#:~:text=return%20an%20error.-,CNI_ARGS,-%3A%20Extra%20arguments%20passed). But having a non-standard constant in CNI plugin is somehow weird (and that's why #435 is still there), so I think users should use this syntax to achieve [this syntax](https://gist.github.com/addreas/92b187332cb100c5935fb235a6620529):
```cjson
{
    "cniVersion": "0.4.0",
    "plugins": [
        {
            //...
            "ipam": {
                "type": "dhcp",
                // ...
                "send": [
                    //...
                    {
                        "option": "host-name", // see man page dhcp-options(5)
                        "fromArg": "K8S_POD_NAME"
                    }
                ]
            }
        }
    ]
}
```

Also, this means the configuration syntax is slightly different from my previous version. Instead of using `value-source`, the parameter is now called `fromArg` which means get the value from CNI arguments.

## Possible future change on the client identifier

And I also truncate `dhcp-client-identifier` to 254 bytes, but a separate issue should be opened for change the identifier syntax. Currently, the identifier looks like `container_id + "/" + netns + "/" + ifname`, but the identifier is not meant to be human readable<sup>1</sup>. But I'm not sure how many users are actively using this feature, and this is a breaking change (though minor), so I think a better way to solve this is open a new issue about it.

## Tech debt

Besides, current DHCP plugin has some tech debt. It's basically modified from d2g/dhcp4 with some copy & paste, and the upstream has not been updated for 3 years. This is not a good news. Also, this means some critical features like `dhcpv6` and sub-option (which is necessary to provide `remote-id` in [RFC4243](https://datatracker.ietf.org/doc/html/rfc4243)) processors. I would recommend<sup>2</sup> using implementation in insomniacslk/dhcp to add support for new features, and simplify code at CNI side. But this should also left as another issue.

## Conclusion

So in conclusion, I would like to end this issue with a PR about a new config syntax, some minor fix, and allowing some fine control over DHCP options. The other issues should be separately dealt with.

The PR will be send tomorrow (since it's nearly 6AM and I need some sleep). Also, I think #663 is important, so if any PR for #663 is ongoing I would manually rebase to that PR.

cc @dcbw @vivekskrishna and welcome for any suggestions!


---

1. Actually in both kea-dhcpv4 and dnsmasq lease file, the identifier looks like mac address. Current representation make those addresses looks like being ascii encoded.
2. In fact, since `remote-id` is impossible to implement in current plugin, I wrote an DHCP proxy using this package to add `remote-id` for `dhcp` ipam plugin.

**Comment by vivekskrishna** [comment]
@SilverBut Nice job. Can you also add support for OptionClasslessRouteFormat option as the code is already present to decode the response for this option. This will also fix #640 .

Also as a long term , we should think about providing the client ID via some pod/deployment annotation. In this way we can make sure that even if pod/deployment moves across worker ndoes and if for some reason , wants to retain its IP,  it is possible as long as it is properly annotated. (ofcourse if replica > 1 it is not much useful)

**Comment by MaxThom** [comment]
Hi,
is it possible to set an option that prevent the dhcp plugin to add a default route? it conflicts with the kubernetes defautl one.
